File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    int[] nums;
    int cnt = 0, len = 0, t;
    public int solution(int[] numbers, int target) {
        nums = numbers;
        len = numbers.length;
        t = target;
        
        // for (int num: nums)
        //     System.out.println(num);
        // System.out.println(cnt);
        // System.out.println(len);
        // System.out.println(t);
        dfs(0, 0);
        
        return cnt;
    }
    public void dfs(int n, int cur){
        if (n == len){
            if (cur == t){
                cnt++;
            }
            return;
        }
        dfs(n + 1, cur + nums[n]);
        dfs(n + 1, cur - nums[n]);
    }
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
end_num = 0
nums = []
cnt = 0


def dfs(cur, depth):
    global end_num
    if depth == len(nums):
        if cur == end_num:
            global cnt
            cnt += 1
        return
    dfs(cur + nums[depth], depth + 1)
    dfs(cur - nums[depth], depth + 1)


def solution(numbers, target):
    global end_num, nums
    end_num = target
    nums = numbers
    answer = 0
    dfs(0, 0)
    return cnt